### PR TITLE
Feature/35226 send email spec key

### PR DESCRIPTION
**Pull request by YRedko**

### CODE REVIEW

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecService.java [hunk 1]
```diff
@@ -20,19 +24,35 @@ public class EmailSpecService extends AbstractRefreshableConfiguration<EmailSpec
...
         }
 
         EmailSpec emailSpec = getConfigurations().get(cfgTenantKey);
         CustomEmailSpec customEmailSpec = customEmailSpecService.getConfigurations().get(cfgTenantKey);
-        return emailSpec.override(customEmailSpec);
+        return Optional.of(emailSpec.override(customEmailSpec));
+    }
+
+    public Optional<EmailTemplateSpec> getEmailTemplateSpec(String tenantKey, String templateKey) {
+        if (getEmailSpec(tenantKey).isEmpty()) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecService.java:43]: use map methods

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecService.java [hunk 2]
```diff
@@ -20,19 +24,35 @@ public class EmailSpecService extends AbstractRefreshableConfiguration<EmailSpec
...
+            return Optional.empty();
+        }
+        List<EmailTemplateSpec> emailTemplateSpecList = getEmailSpec(tenantKey).get().getEmails();
+        return emailTemplateSpecList.stream().filter(it -> it.getTemplateKey().equals(templateKey)).findFirst();
+    }
+
+    public Optional<EmailTemplateSpec> getEmailTemplateSpecByPath(String tenantKey, String templatePath) {
+        if (getEmailSpec(tenantKey).isEmpty()) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecService.java:51]: use Optional methods

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/TenantEmailTemplateService.java [hunk 3]
```diff
@@ -32,30 +35,53 @@ public class TenantEmailTemplateService implements RefreshableConfiguration {
...
+    @LoggingAspectConfig(resultDetails = false)
+    public String getEmailTemplateByKey(TenantKey tenantKey, String templateKey, String locale) {
+        Optional<EmailTemplateSpec> emailTemplateSpec = emailSpecService.getEmailTemplateSpec(tenantKey.getValue(), templateKey);
+        if (emailTemplateSpec.isPresent()) {
+            String templatePath = emailTemplateSpec.get().getTemplatePath();
+            Optional<String> emailTemplateOptional =  getTemplateOverrideable(templatePath);
+
+            return emailTemplateOptional.orElseGet(() -> getEmailTemplate(tenantKey.getValue(), templateKey, locale));
+        }
+        return getEmailTemplate(tenantKey.getValue(), templateKey, locale);
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/TenantEmailTemplateService.java:66]: emailSpecService.getEmailTemplateSpec(tenantKey.getValue(), templateKey)
  .map(EmailTemplateSpec::getTemplatePath)
  .map(this::getTemplateOverrideable)
  .orElseGet(() -> getEmailTemplate(tenantKey.getValue(), templateKey, locale));

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java [hunk 4]
```diff
@@ -312,19 +316,21 @@ private void initAndSendEmail(TenantKey tenantKey,
...
+            String emailTemplate = tenantEmailTemplateService.getEmailTemplateByKey(tenantKey, templateName, locale.getLanguage());
+            String processedContent = emailTemplateService.processEmailTemplate(tenantKey.getValue(), emailTemplate, objectModel, locale.getLanguage());
 
             try {
                 tenantContextHolder.getPrivilegedContext().setTenant(new PlainTenant(tenantKey));
 
-                Template mailTemplate = new Template(templateKey, emailTemplate, freeMarkerConfiguration);
-                String content = FreeMarkerTemplateUtils.processTemplateIntoString(mailTemplate, objectModel);
                 MailParams mailParams = resolve(subject, from, templateName, locale, objectModel);
+
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java:326]: add resolve by spec

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java [hunk 5]
```diff
@@ -312,19 +316,21 @@ private void initAndSendEmail(TenantKey tenantKey,
...
+                Template subjectTemplate = new Template(UUID.randomUUID().toString(), mailParams.getSubject(), freeMarkerConfiguration);
+                String renderedSubject = FreeMarkerTemplateUtils.processTemplateIntoString(subjectTemplate, objectModel);
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java:328]: replace to render in mailParams

**Review by sergeysenja1992 — APPROVED** [review]

**Review by alexmoldavsky — APPROVED** [review]

File: src/test/java/com/icthh/xm/tmf/ms/communication/service/mail/EmailTemplateServiceUnitTest.java [hunk 6]
```diff
@@ -83,6 +84,8 @@ public class EmailTemplateServiceUnitTest {
     @Mock
     private CommonConfigRepository commonConfigRepository;
 
+    private static final String LANG_KEY = "en";
```
Comment (monassfaro) [inline src/test/java/com/icthh/xm/tmf/ms/communication/service/mail/EmailTemplateServiceUnitTest.java:87]: We already have a language constant - use Constants.DEFAULT_LANGUAGE

File: src/test/java/com/icthh/xm/tmf/ms/communication/web/rest/errors/ExceptionTranslatorTestController.java [hunk 7]
```diff
@@ -60,7 +60,7 @@ public void internalServerError() {
 
     @PostMapping("/test/render-template-exception")
     public void renderTemplateException() {
-        throw new RenderTemplateException("render failed", "${{test", Map.of("test", "test"));
+        throw new RenderTemplateException("render failed", "${{test", Map.of("test", "test"), "en");
```
Comment (monassfaro) [inline src/test/java/com/icthh/xm/tmf/ms/communication/web/rest/errors/ExceptionTranslatorTestController.java:63]: The same as above - use Constants.DEFAULT_LANGUAGE

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java [hunk 8]
```diff
@@ -376,6 +373,25 @@ private MailParams resolve(String subject,
...
+        Optional<EmailTemplateSpec> templateSpec = emailSpecService.getEmailTemplateSpec(tenantKey, templateKey);
+
+        if (templateSpec.isPresent()) {
+            EmailTemplateSpec emailTemplateSpec = templateSpec.get();
+            log.info("resolve by spec: for templateKey: {} found templateSubject: {}",
+                emailTemplateSpec.getTemplateKey(), emailTemplateSpec.getSubjectTemplate());
+
+            if (StringUtils.isBlank(mailParams.getSubject())) {
+                Map<String, String> langToSubjectMap = emailTemplateSpec.getSubjectTemplate();
+                String i18nSubject = langToSubjectMap.getOrDefault(lang,langToSubjectMap.get("en"));
```
Comment (monassfaro) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java:386]: Use - Constants.DEFAULT_LANGUAGE

**Pull request merged**